llvm/include/llvm/SandboxIR/SandboxIR.h

```diff
@@ -1020,10 +1021,61 @@ class ConstantPointerNull final : public Constant {
 #endif
 };
 
-// TODO: Inherit from UndefValue.
-class PoisonValue final : public Constant {
+// TODO: Inherit from ConstantData.
+class UndefValue : public Constant {
+protected:
+  UndefValue(llvm::UndefValue *C, Context &Ctx)
+      : Constant(ClassID::UndefValue, C, Ctx) {}
+  UndefValue(ClassID ID, llvm::Constant *C, Context &Ctx)
+      : Constant(ID, C, Ctx) {}
+  friend class Context; // For constructor.
+
+public:
+  /// Static factory methods - Return an 'undef' object of the specified type.
+  static UndefValue *get(Type *T);
+
+  /// If this Undef has array or vector type, return a undef with the right
+  /// element type.
+  UndefValue *getSequentialElement() const;
+
+  /// If this undef has struct type, return a undef with the right element type
+  /// for the specified element.
+  UndefValue *getStructElement(unsigned Elt) const;
+
+  /// Return an undef of the right value for the specified GEP index if we can,
+  /// otherwise return null (e.g. if C is a ConstantExpr).
+  UndefValue *getElementValue(Constant *C) const;
+
+  /// Return an undef of the right value for the specified GEP index.
+  UndefValue *getElementValue(unsigned Idx) const;
+
+  /// Return the number of elements in the array, vector, or struct.
+  unsigned getNumElements() const {
+    return cast<llvm::UndefValue>(Val)->getNumElements();
+  }
+
+  /// For isa/dyn_cast.
+  static bool classof(const sandboxir::Value *From) {
+    return From->getSubclassID() == ClassID::UndefValue ||
+           From->getSubclassID() == ClassID::PoisonValue;
```

Comment:
isa<PoisonValue>(Undef) == true?

Reply:
Hmm did you mean `return From->getSubclassID() == ClassID::UndefValue || isa<PoisonValue(From)` ?
I would not prefer this because because it hides the fact that PoisonValue corresponds to just a single `ClassID::PoisonValue`.

Hmm or  did you mean checking the llvm-side with something like `isa<PoisonValue>(From->Val)` ?

The reason why I don't like this is that it looks like it is slightly slower: Because you need to get `From->Val` and then run `classof()` which does the comparison `ID == UndefValue || ID == PoisonValue`.

Reply:
You modelled it correctly. Poison inherits from Undef, but an undef value/instance should not be poison.

Reply:
```
 static bool classof(const sandboxir::Value *From) {
    return From->getSubclassID() == ClassID::UndefValue;
}
```

should work.

Reply:
Yes, `isa<>` works  because it takes care of upcasting. But `classof()` would be wrong, it should return true if `From` is a `PoisonValue`.

Reply:
I believe the current classof UndefValue is wrong. It should only test for ClassId::UndefValue.

```
isa<PoisonValue>(Undef)
```
should/must be false.

Reply:
classof() should return true when `From` is either the current class type (that is `UndefValue`) or its sub-classes (that is `PoisonValue`). This is implemented the same way as in `llvm/IR/Constants.h`.

Reply:
> isa<PoisonValue>(Undef) should/must be false.

We are already testing that: ` EXPECT_FALSE(isa<sandboxir::PoisonValue>(Val: Undef)); // Undef is not Poison  `